File: DependencyInjectionTesting/DependencyInjection.FSharp.Tests/CreateControllerShould.fs
module DependencyInjection.FSharp.Tests.Tests.DependencyInjectionTests

open System
open System.Collections.Generic
open Application.Api
open FsUnit.Xunit
open DependencyInjection.FSharp.Tests.Helpers
open DependencyInjection.FSharp.Tests.Helpers.DependencyInjectionHelper
open Xunit

module Config =
    let basicConfig =
        [
            kvp "Config:Key" "null"
        ]

open Config

let controllersToTest: IEnumerable<Object []> =
        typeof<Controllers.BaseController>
        |> ControllerFinder.findControllers
        |> Seq.map (fun t -> t :> Object)
        |> Seq.map Array.singleton

[<Theory>]
[<MemberData("controllersToTest")>]
let ``Successfully create the controllers with their dependencies`` controllerType =
    List.concat [basicConfig]
    |> registerServices
    |> addController controllerType
    |> buildServiceProvider
    |> getController controllerType
    |> should not' (be Null)

File: DependencyInjectionTesting/DependencyInjection.CSharp.Tests/CreateControllerShould.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Application.Api;
using Application.Api.Controllers;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Xunit;

namespace DependencyInjection.CSharp.Tests
{
    public class CreateControllerShould
    {
        private readonly IServiceCollection _services;

        public CreateControllerShould()
        {
            _services = RegisterServices();
        }

        [Theory]
        [MemberData(nameof(Controllers))]
        public void SuccessfullyLoadDependencies(Type controllerType)
        {
            _services.AddTransient(controllerType);
            var provider = _services.BuildServiceProvider();

            var controller = provider.GetRequiredService(controllerType);

            Assert.NotNull(controller);
        }

        public static IEnumerable<object[]> Controllers()
        {
            return ControllerFinder
                .FindControllers(typeof(BaseController))
                .Select(x => new[] {x});
        }

        private static IServiceCollection RegisterServices()
        {
            var config = CreateConfiguration();
            var startup = new Startup(config);

            return CreateServiceCollection(startup);
        }

        private static IConfigurationRoot CreateConfiguration()
        {
            var configValues = new[]
            {
                new KeyValuePair<string, string>("Config:Key", "null")
            };

            return new ConfigurationBuilder()
                .AddInMemoryCollection(configValues)
                .Build();
        }

        private static ServiceCollection CreateServiceCollection(Startup startup)
        {
            var services = new ServiceCollection();

            //Add services provided by default
            services.AddLogging(x => x.AddConsole());
            //Add application specific services
            startup.ConfigureServices(services);

            return services;
        }
    }
}